tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/impl/StableTopologicalSort.scala

```diff
@@ -0,0 +1,66 @@
+package at.forsyte.apalache.tla.lir.transformations.impl
+
+/**
+ * An implementation of a topological sort following <a href="https://en.wikipedia.org/wiki/Topological_sorting">Kahn's algorithm</a>.
+ * Our sorting is stable, that is, when two graph nodes belong to the same layer, they are ordered by the original ordering.
+ * (We are assuming that the input list contains no duplicates.)
+ *
+ * @tparam T the type of graph vertices
+ * @author Igor Konnov
+ */
+class StableTopologicalSort[T]() {
+  type Edges = Map[T, Set[T]]
+
+  /**
+   * Sort the elements of the list `unsorted` according to the dependencies that are stored in the incoming edges.
+   * We are assume that `inEdges` contain values for all elements of `unsorted`.
+   *
+   * @param inEdges  a set of incoming edges per every element of unsorted (may be empty)
+   * @param unsorted a list of nodes
+   * @return either Left(sorted) that contains the sorted nodes, or Right(nodes) that contains a subgraph with a cycle inside.
+   */
+  def toposort(inEdges: Edges, unsorted: Seq[T]): Either[List[T], Set[T]] = {
+    require(inEdges.keySet == unsorted.toSet)
+
+    // save the unsorted order to guarantee stability
+    val originalOrder = unsorted.zipWithIndex.foldLeft(Map.empty[T, Int]) { case (m, (e, i)) => m + (e -> i) }
+
+    // Use Kahn's algorithm to sort the declarations in a topological order:
+    // https://en.wikipedia.org/wiki/Topological_sorting
+
+    // the list of sorted nodes
+    var sorted = List.empty[T]
+    // the edges that have not been closed in the graph yet
+    var edges = inEdges
+    // the list of nodes that do not have incoming edges
+    var sinks = List.empty[T]
+
+    def updateSinks(): Unit = {
+      val (sinkEdges, otherEdges) = edges.partition(_._2.isEmpty)
+      // since the sinks have no incoming edges, we can sort them by the original order
+      sinks = sinkEdges.keys.toList.sortWith { (l, r) =>
+        originalOrder(l) < originalOrder(r)
+      }
+      edges = otherEdges
+    }
+
+    // initialize sinks with the nodes that have no incoming edges
+    updateSinks()
+    while (sinks.nonEmpty) {
+      sorted ++= sinks
+      val toRemove = sinks.toSet
+      // remove all incoming edges that contain one of the sinks as a source
+      edges = edges.transform { (_, callers) =>
+        callers -- toRemove
+      }
+      // recompute the sinks and edges
+      updateSinks()
+    }
```

Comment:
I could be wrong, but this doesn't look to follow Kahn's algorithm. From you reference, Kahn's process once sink node at a time, and doesn't need to recompute the sinks and edges, or have to reprocess the edges in each iteration to remove sinks (like we do here on line 54). Since you say this uses Kahn's algorithm, I think it would be useful to add a line explaining why you opted to make these substantive alterations. (I'd also be curious myself ;)).

Reply:
True. The idea is the same, but I am building declaration layer-by-layer. This allows us to sort the definitions inside every layer by their original ordering. To me it is almost the same algorithm, just written in a more functional way. The pseudocode in Wikipedia is quite non-deterministic, and one of its executions produces the same list as in our case. My code is more predictable.

Reply:
That's a very good reason!  You explanatory comments are perfect :+1:  Thanks :)